Source/dom/components/Library.js

```diff
@@ -0,0 +1,141 @@
+import { WrappedObject, DefinedPropertiesKey } from '../WrappedObject'
+import { Document } from './Document'
+import { toArray, getURLFromPath } from '../utils'
+import { Types } from '../enums'
+import { Factory } from '../Factory'
+import { wrapObject } from '../wrapNativeObject'
+import { ShareableObject } from './ShareableObject'
+
+const AddStatus = {
+  0: 'ok',
+  1: 'the library has already been added',
+  2: 'the document is not in the new JSON format',
+  3: 'there was a problem reading the asset library file',
+}
+
+/**
+ * A Sketch Library.
+ */
+export class Library extends WrappedObject {
+  constructor(library = {}) {
+    if (!library.sketchObject) {
+      throw new Error('Cannot create a new Library directly')
+    }
+
+    super(library)
+  }
+
+  static getLibraries() {
+    const libraryController = AppController.sharedInstance().librariesController()
+    return toArray(libraryController.libraries()).map(
+      Library.fromNative.bind(Library)
+    )
+  }
+
+  static createLibraryFromDocument(document, path) {
+    if (typeof document === 'string' && !path) {
+      /* eslint-disable no-param-reassign */
+      path = document
+      document = undefined
+      /* eslint-enable */
+    }
+
+    const libUrl = getURLFromPath(path)
+
+    if (document) {
+      const wrappedDocument = wrapObject(document)
+      wrappedDocument.save(path)
+    }
+
+    const libraryController = AppController.sharedInstance().librariesController()
+    const status = libraryController.addAssetLibraryAtURL(libUrl)
+
+    if (status !== 0) {
+      throw new Error(`Error while adding the library: ${AddStatus[status]}.`)
+    }
+
+    // refresh the UI
+    libraryController.notifyLibraryChange(null)
+
+    const lib = toArray(libraryController.userLibraries()).find(l =>
+      l.locationOnDisk().isEqual(libUrl)
+    )
+
+    if (!lib) {
+      throw new Error('could not find the added library')
+    }
+
+    return Library.fromNative(lib)
+  }
+
+  getDocument() {
+    if (!this._object.document() && !this._object.loadSynchronously()) {
+      throw new Error(`could not get the document: ${this._object.status}`)
+    }
+    return Document.fromNative(this._object.document())
+  }
+
+  getSymbolReferences() {
+    try {
+      const document = this.getDocument()
+
+      return document
+        .getSymbols()
```

Comment:
Also, depending on what's being done with these references we may want the symbol master to come from a local document rather than the library. For instance; lets say you have a library with a red symbol and you import that symbol into your document. You then change the symbol in the library to be blue, but don't update your document. When you go to insert another instance of that symbol into your document you expect the symbol preview to be red not blue. However, if the reference you are using is pointing to the library's version (as it is here) the preview would be blue.

It might make more sense to make this `symbolReferencesForDocument` or conversely have a `symbolReferencesForLibrary` on the document. It would also be worth taking a look at `MSForeignObjectCollector` as this will do most of this matching for you - I think this uses blocks so, if it helps, we could always add a helper method on the obj-C side.

Reply:
> lets say you have a library with a red symbol and you import that symbol into your document. You then change the symbol in the library to be blue, but don't update your document. When you go to insert another instance of that symbol into your document you expect the symbol preview to be red not blue. However, if the reference you are using is pointing to the library's version (as it is here) the preview would be blue.

That's actually what I would expect. If you get a hold of a library, import one of its symbol, I'd expect the symbol to be the one of th

Reply:
Not sure what you're saying here?

Reply:
I would expect that when I'm accessing a library and importing one of its symbols, it's importing the library's version and not the existing document's version

Reply:
Hmm - that's not how Sketch works at the moment (by design).

Reply:
Hum I see. I can understand that. Will change.

Is there a way to programmatically bring the changes from an updated library?

Reply:
You want to use `syncForeignObject:withMaster:fromLibrary:` on `MSAssetLibraryController`.

If you wanted I think you could pass in `nil` for the second parameter and (assuming it's the correct library).

This will do the right thing in terms of importing any missing, nested symbols but if the symbol has already been imported it won't update them - you'd have to call `syncForeignObject:withMaster:fromLibrary:` on them to achieve that.